create proposal for a add_transformation function

nikosbosse opened the pull request:
I just implemented a first proposal for #270. 

I think one question is whether we want to call the extra column 'scale' or something else. 'transformation'? I like the word scale, but also "log scale" is potentially ambiguous. 

I haven't done any checking here (which one could do), but I though it would be bit of an overkill since it's an extremely simple function and running eg `check_forecasts()` can take some while.

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+#' @title Add transformations
+#'
+#' @description Add transformations of the forecasts and observations for
```
I think you can give a little more detail here as this is extremely sparse and doesn't really make people want to check the reference.

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+#'
+#' @inheritParams score
+#' @param fun A function used to transform both true values and predictions
+#' @param label A string for the newly created 'scale' column to denote the
```
If you are renaming these columns how do you expect this to work with score? I think you should be showing this workflow here as an example? 

I assumed the logical thing to do was transform the existing columns so you can score them and keep a copy as natural scale?

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+#' out the linked reference.
+#'
+#' @inheritParams score
+#' @param fun A function used to transform both true values and predictions
```
I think a lot of the value for this comes from adding a few default options along with justification. The aim of the package is to be helpful right - at the moment this is such thin wrapper it isn't clear it adds much.

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+
+#' @keywords check-forecasts
+#' @examples
+#' add_transformation(example_quantile)
```
I think this needs more full-some examples

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+#' add_transformation(example_quantile)
+
+add_transformation <- function(data,
+                               fun = function(x) {log(x + 1)},
```
this default isn't documented or flagged in any way

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+add_transformation <- function(data,
+                               fun = function(x) {log(x + 1)},
+                               label = "log",
+                               ...) {
```
no doc of ...

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,48 @@
+  transformed_data[, prediction := fun(prediction, ...)]
+  transformed_data[, true_value := fun(true_value, ...)]
+
+  out <- rbind(data, transformed_data)
```
I don't really like this combining datasets feature. How are you thinking this would work in a workflow? 

This was the use case I was imagining:

```r
forecasts |>
  scale_forecasts(option = "log") |>
  check_forecasts() |>
  score() |>
  summarise_score(by = "model")
# Forecasts and observations have been transformed to the log scale. This changes the interpretation to a relative score in most cases. See: for details
# blah blah check forecasts output. 
```

What was the workflow you imagined?

seabbs replied:
I imagine its scoring all together and the summarising across scale? This could be a nice approach but needs documentation and some justification!

seabbs commented on tests/testthat/test-convenience-functions.R:
```diff
@@ -0,0 +1,12 @@
+
+  predictions <- example_quantile$prediction
+
+  one <- suppressWarnings(add_transformation(example_quantile))
```
where are the warnings coming from?

seabbs reviewed (commented):
I think this needs some thought on what the workflow would look like but the idea is nice.

seabbs reviewed (commented):
This looks nice. I was a little confused by the workflow you were proposing so more docs for that I think (though if its as I imagine it seems nice enough). Also just need more docs in general and adding some default options with verbose messaging of what they mean seems like it would be very useful. 

How do you imagine this function being used in serial (so I want to have a series of transformations). At the moment I think it will just break but could easily code around this.

nikosbosse commented:
Ah yes, I think your suggestion makes probably more sense than mine if you have more than one transformation. The workflow I had in mind was something like 

```
forecasts |>
  add_transformation(option = "log") |>
  score() |>
  summarise_score(by = c("model", "scale"))
```
with results something like 

| model | wis | scale |
| ------ | --- | ------ |
| A.        | 2.   | natural | 
| A.        | 0.1  | log | 
| B.        | 3.   | natural |
| B.        | 0.3 | log |

So it would automatically just add the log scaled data to your data.frame and add a separate column. With your alternative suggestion the equivalent workflow would be something like 

```
forecasts |>
  mutate(scale = "natural") |>
  rbind(
      scale_forecasts(option = "log") |>
      mutate(scale = "log")) |>
  score() |>
  summarise_score(by = c("model", "scale"))
```


(the additional suggestion in your workflow to enable `data |>  check_forecasts() |>  score()` would require an additional change to `check_forecasts()` that we probably should deal with separately)

seabbs commented:
I like your workflow as its certainly less clunky but yes worry about what happens when you want multiple transformations. Perhaps the solution is to add some code logic t o`add_transformation` so that it looks for a `scale` variable and if it finds it filters for `natural` and applies everything there then appends to the end vs doing the current logic if it doesn't find it? 

The alternative would be making it be called once and enabling passing a list of transformations but I am not a big fan of that as it hides quite a lot of code internally for small gain.. 

I hadn't realised `check_forecasts` wasn't pipeable - I guess it doesn't totally need to be but it would be nice if it were. Solutions are likely some kind of s3 class I would guess.

nikosbosse commented:
Trying to summarise the discussion: 
we want to 
- implement the version with `add_transformation(option = "log")` 
    - have a way to do that multiple times, checking whether there is already a "scale" column
- have your proposed function that transforms without adding as well?  
- have default options? 
    - problem with that one e.g. for the log would be that one would need to pass in the offset as an extra argument
    - and for population that would require a very specific column with a name that's appropriate in epi, but maybe not in other instances
- have the possibility to pass in a custom function? I
    -  think that would be nice
- have both options, i.e. some prespecified choicse + custom functions?

seabbs commented:
I think we: 

> implement the version with add_transformation(option = "log")
have a way to do that multiple times, checking whether there is already a "scale" column

Want this as its nice and clear for one but feel we need to support multiple. 

> have your proposed function that transforms without adding as well?

No I don't think we need this as well.

> have default options?
problem with that one e.g. for the log would be that one would need to pass in the offset as an extra argument
and for population that would require a very specific column with a name that's appropriate in epi, but maybe not in other instances

I think we want defaults as lots of users won't know what to use. For the log we could just set the offset and for more advanced use expect users to specify manually (we can clearly document this). 

For population adjustments its really just a normalisation using an offset which is common everywhere if we label it as that and say in the docs its using it for population would mean x might work? 

> have the possibility to pass in a custom function? I
think that would be nice

Agree it definitely needs this and can just have this in the first instance (though has default options aren't that much work I would suggest we do both).

sbfnk commented:
> implement the version with add_transformation(option = "log")
> have a way to do that multiple times, checking whether there is already a "scale" column

Are the transformation always targeting the natural values? Or might we want to apply a scaling to the scaled values? The former makes more sense to me overall (as anything could be passed in a custom function) but it might be a bit complicated having to communicate to the user that what the operation does depends on whether a `scale` column is present and contains values "natural". Alternatively the log+1 could be implemented as e.g.

```
forecasts |>
  transform(operation = sum,  offset = 1) |>
  transform(operation = log) |>
  score() |>
  summarise_score(by = c("model"))
```

which would remove the need to name the transform and then filter out columns values (I think this is your option (a), @nikosbosse).

seabbs commented:
@sbfnk for your suggestion you would lose the ability to score natural and transformed forecasts at once as in @nikosbosse approach + wouldn't be able stack transformations without the addition of code. On the plus side the code is much simpler, the workflow for a single transform is cleaner, and you don't need to bother labelling things which is a pain/ 

It is similar in spirit to my suggestion here: https://github.com/epiforecasts/scoringutils/pull/271#discussion_r1119869202

Having seen @nikosbosse suggest use case I do like the idea of scoring multiple transformations at once and having functionality to make this easier. Not entirely convinced its worth the complexity trade-offs you highlight though. 

Default options: I really like the idea of default options (like those supplied in the `scale_` `ggplot2` functions) as its a nice hand holder for less experienced users and they can easily be documented (i.e in the examples). That said it really isn't much code to add these so perhaps just having examples and links to reading more is all that is needed.

nikosbosse commented:
So I updated the suggested code a bit: 
- For now I decided against dedicated default options. Mainly because `option = "log"` doesn't seem that much easier to me than `fun = log`. Open to being convinced otherwise. 
- For `fun = log` I added some "clever" code to check whether there are any zero values. If there are, then it automatically applies an offset and complains to the user with a warning. You can pass an argument `offset = 1` or whatever offset you like to the function. I think the workflow is well documented and intuitive, but I imagine this to be somewhat controversial :D 
- it has an argument `add = TRUE` to toggle whether you want to transform the existing data, or whether you want to add the transformed data to your existing data. This also works multiple times and there is an example for it

What do you think? 



Some workflow examples: 

```
#' library(magrittr) #(pipe operator)
#' # replace negative values with zero
#' transformed <- example_quantile %>%
#'   transform_forecasts(fun = function(x) {pmax(0, x)}, add = FALSE)
#'
#' # add log transformed forecasts (produces a warning as some values are zero)
#' transform_forecasts(transformed)
#'
#' # specifying an offset manually for the log transformation removes the warning
#' transform_forecasts(transformed, offset = 1)
#'
#' # adding multiple transformations
#' transformed %>%
#'   transform_forecasts(offset = 1) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
#'   score() %>%
#'   summarise_scores(by = c("model", "scale"))
```

sbfnk commented:
Alternative would be to create a small function `log_shift` or similar that removes the need for filtering through dot arguments, hiding default options and checking whether any function is a specific premitive?

```r
log_shift <- function(x, offset = 1, ...) {
  return(log(x + offset, ...))
}
```

This could then be the default argument for `transform_forecasts` if you really wanted a default argument.

> it has an argument add = TRUE to toggle whether you want to transform the existing data, or whether you want to add the transformed data to your existing data. This also works multiple times and there is an example for it

I'd call it `append` or similar - `add` to me implies numerical addition as in e.g. `magrittr::add`

nikosbosse commented:
Maybe the best idea would be to just have an option `offset = 0` (or `shift = 0`?) that just adds one to everything. It could then automatically do that for the log (or not do it and just complain / alert the user, which I guess is the better option). 
With this we wouldn't have to change the log function which I think is good, but there would also be a simple argument that allows the user to apply an offset.

seabbs commented:
I still prefer @sbfnk's suggestion of a custom transformation function vs giving special design consideration in this function for the log transform but your call.

seabbs commented:
@nikosbosse any help needed to unstick this?

nikosbosse commented:
No real blockers, I was waiting on more feedback from Le Big Boss. I'll implement the `log_shift()` version :)

nikosbosse commented:
You had cast your vote :) 
Implemented a new suggestion. Current draft exports a function `log_shift()` that allows the user to add an offset and also truncate negative values in one function call. Default offset is 0.

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+#' library(magrittr) # pipe operator
+#' example_quantile %>%
+#'   # truncate all negative values for both log and sqrt
+#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE) %>%
```
out of date?

nikosbosse replied:
It's needed as the sqrt function doesn't have a truncate option. That part is maybe a bit confusing... The truncate option is part of `log_shift()`, not of `transform_forecasts()`

nikosbosse replied:
I changed the example to this: 
```
#' example_quantile %>%
#'   transform_forecasts(offset = 1, truncate = TRUE) %>%
#'   # manually truncate all negative values before sqrt
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
#'   score() %>%
#'   summarise_scores(by = c("model", "scale"))
```

maybe that's a bit more intuitive (it also helped me catch an error where if you chose append = FALSE it would override previous transformations)

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+#' the argument `label`).
+#' @param label A string for the newly created 'scale' column to denote the
+#' newly transformed values. Only relevant if `append = TRUE`.
+#' @param ... Additional parameters to pass to the function you supplied. If
```
out of date

nikosbosse replied:
thanks!

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+#'
+#' @inheritParams score
+#' @param fun A function used to transform both true values and predictions.
+#' If `fun = log` it automatically checks internally whether there are any
```
out of date

nikosbosse replied:
fixed, thanks

seabbs commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+                                label = "log",
+                                ...) {
+
+  if (identical(fun, sqrt)) {
```
not a huge fan of building in these kind of function-specific exceptions

nikosbosse replied:
hm yeah - can also get rid of it. Only goal was to provide more useful error messages, but you're right, probably not necessary

seabbs reviewed (commented):
Looks good to me. Seems like a nice workflow. Just flagged that the docs are currently out of date. 

Not for this PR by a vignette on forecast transformation seems like a nice and useful idea.

nikosbosse commented:
@sbfnk @seabbs happy to merge?

sbfnk commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+    x <- pmax(0, x)
+  }
+
+  if (any (x < 0, na.rm = TRUE)) {
```
do as lint tells you and remove the space?

sbfnk commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+  }
+  log(x + offset, base = base)
+}
+
```
"Trailing blank lines are superfluous"

sbfnk commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+#' @param x vector of input values to be transformed
+#' @param offset number to add to the input value before taking the natural
+#' logarithm
+#' @param truncate whether to truncate negative values to 0. Default is FALSE.
```
I don't know what this means before looking at the code. Does the truncation happen before or after the transform, and why would one want to truncate?

sbfnk commented on R/convenience-functions.R:
```diff
@@ -0,0 +1,194 @@
+log_shift <- function(
+    x,
+    offset = 0,
+    truncate = FALSE,
```
Do we really want this option here? I'd think it's better for the log function to throw an error and the user to pre-process the forecasts/data so that there are no negatives.

If sticking with it I would suggest to make this more self-explanatory, i.e. calling the option `negative_to_zero` or something like that (although given that we might be adding an offset anyway I'm not clear why we'd choose zero over anything else).

nikosbosse replied:
I think the reason to have that truncation is that the alternative (at least staying in the same workflow) is a bit more clunky. You would have to use something like `transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural")`

See e.g. these examples: 

```
#' example_quantile %>%
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") 
```

```
#' # adding multiple transformations
#' library(magrittr) # pipe operator
#' example_quantile %>%
#'   transform_forecasts(offset = 1, truncate = TRUE) %>%
#'   # manually truncate all negative values before applying sqrt
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural") %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
#'   score() %>%
#'   summarise_scores(by = c("model", "scale"))
```

(the part with the label = natural here is only because it's transforming the original while something else has already been computed, which maybe is a stupid thing to do?)

nikosbosse replied:
Looking at it again I should definitely simplify the second example...

sbfnk replied:
> I think the reason to have that truncation is that the alternative (at least staying in the same workflow) is a bit more clunky. You would have to use something like `transform_forecasts(fun = function(x) pmax(0, x), append = FALSE, label = "natural")`

Wouldn't it be more natural to do something like:

```r
example_quantile %>%
  dplyr::filter(true_value >= 0) %>%
  transform_forecasts(fun = sqrt, label = "sqrt") 
```

or 

```r
example_quantile %>%
  dplyr::mutate(true_value = dplyr::ifelse(true_value < 0, 0, true_value)) %>%
  transform_forecasts(fun = sqrt, label = "sqrt") 
```

i.e. fix the problem in the data rather than letting the transformation function take care of it?

sbfnk reviewed (commented):
Nearly there. Just one more question.

nikosbosse commented:
@seabbs do you have feelings about the `negative_to_zero` argument and whether or not it should be kept? 
Pro: it's a bit more convenient for users who know what they're doing (as they can replace forecasts and observations in one go)
Con: it's a bit cleaner to force users to make the changes manually 

Current version has the examples replaced to make everything a bit easier to understand:

```
#' library(magrittr) # pipe operator
#'
#' # add log transformed forecasts (produces a warning as some values are zero)
#' # negative values need to be handled (here by replacing them with 0)
#' example_quantile %>%
#'   .[, true_value := ifelse(true_value < 0, 0, true_value)] %>%
#'   transform_forecasts(append = FALSE)
#'
#' # alternatively:
#' transform_forecasts(example_quantile, negative_to_zero = TRUE, append = FALSE)
#'
#' # specifying an offset manually for the log transformation removes the warning
#' example_quantile %>%
#'   .[, true_value := ifelse(true_value < 0, 0, true_value)] %>%
#'   transform_forecasts(offset = 1, append = FALSE)
#'
#' # truncating forecasts manually before sqrt
#' example_quantile %>%
#'   .[, true_value := ifelse(true_value < 0, 0, true_value)] %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt")
#'
#' # alternatively, this achieves the same
#' example_quantile %>%
#'   transform_forecasts(fun = function(x) pmax(0, x), append = FALSE) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt")
#'
#' # adding multiple transformations
#' library(magrittr) # pipe operator
#' example_quantile %>%
#'   .[, true_value := ifelse(true_value < 0, 0, true_value)] %>%
#'   transform_forecasts(offset = 1) %>%
#'   transform_forecasts(fun = sqrt, label = "sqrt") %>%
#'   score() %>%
#'   summarise_scores(by = c("model", "scale"))
```

nikosbosse commented:
@Bisaloo do you have thoughts? :) democratic development

seabbs commented:
I think if we are keeping it then it should be its own helper function. I also think this is could be true for offsetting.

nikosbosse commented:
> I think if we are keeping it then it should be its own helper function. I also think this is could be true for offsetting.

Yes definitely - it's currently an argument of the `log_shift()` helper function (maybe I should adapt the examples again to make that clear). 

the `transform_forecasts()` function currently only has the following arguments: 

```
transform_forecasts <- function(data,
                                fun = log_shift,
                                append = TRUE,
                                label = "log",
                                ...)
```

and `log_shift()` has 

```
log_shift <- function(x,
                      offset = 0,
                      negative_to_zero = FALSE,
                      base = exp(1))
```

seabbs commented:
> - it's currently an argument of the 

Yes, so not an argument but a helper function on its own.

seabbs commented:
A helper independent of transform forecasts

nikosbosse commented:
I merged a slightly different version without a `negative_to_zero` argument in `log_shift()` now. Could still decide to introduce that argument later if we like.
